python/ray/serve/_private/deployment_state.py

```diff
@@ -2248,20 +2248,64 @@ def _check_and_update_replicas(self):
                     del self.replica_average_ongoing_requests[replica.replica_tag]
 
     def _stop_replicas_on_draining_nodes(self):
-        draining_nodes = self._cluster_node_info_cache.get_draining_node_ids()
+        draining_node_deadlines = self._cluster_node_info_cache.get_draining_node_ids()
         for replica in self._replicas.pop(
-            states=[ReplicaState.UPDATING, ReplicaState.RUNNING]
+            states=[ReplicaState.UPDATING, ReplicaState.RUNNING, ReplicaState.STARTING]
         ):
-            if replica.actor_node_id in draining_nodes:
-                state = replica._actor_details.state
+            if replica.actor_node_id in draining_node_deadlines:
+                self._replicas.add(ReplicaState.DRAINING, replica)
+            else:
+                self._replicas.add(replica.actor_details.state, replica)
+
+        # Stop replicas whose deadline is up
+        for replica in self._replicas.pop(states=[ReplicaState.DRAINING]):
+            current_timestamp_ms = time.time() * 1000
+            timeout_ms = replica._actor.graceful_shutdown_timeout_s * 1000
+            if (
+                replica.actor_node_id in draining_node_deadlines
+                and draining_node_deadlines[replica.actor_node_id] > 0
+                and current_timestamp_ms
+                >= draining_node_deadlines[replica.actor_node_id] - timeout_ms
+            ):
                 logger.info(
-                    f"Stopping replica {replica.replica_tag} (currently {state}) "
+                    f"Stopping replica {replica.replica_tag} "
                     f"of deployment '{self.deployment_name}' in application "
                     f"'{self.app_name}' on draining node {replica.actor_node_id}."
                 )
                 self._stop_replica(replica, graceful_stop=True)
             else:
-                self._replicas.add(replica.actor_details.state, replica)
+                self._replicas.add(ReplicaState.DRAINING, replica)
+
+        # Stop excess DRAINING replicas when new "replacement" replicas
+        # have transitioned to RUNNING.
+        num_running = self._replicas.count(states=[ReplicaState.RUNNING])
+        num_draining = self._replicas.count(states=[ReplicaState.DRAINING])
+        num_excess = num_running + num_draining - self._target_state.target_num_replicas
+
+        draining_replicas = self._replicas.pop(states=[ReplicaState.DRAINING])
+
+        # Greedily choose replicas that have the earliest deadline
```

Comment:
Good thinking! I think it'd be easier to follow the logic if you abstracted the logic to get the which replicas to stop into a helper method. i.e.,

```python
def _choose_draining_replicas_to_stop(replicas: List[Replica], deadlines: Dict[str, int], num_to_stop: int) -> List[Replica]:
```

Reply:
Thanks for the suggestion, done!